### ISSUE TITLE
Tarball includes Linux object files

### ISSUE THREAD

**Issue by cpcallen** [mono/mono#14650]
(Edited by @lambdageek, original bug report below)

The following object files (Linux 64-bit ELF, afaict) are included in our tarball, which causes Homebrew to build a bad libmono-native:
```
$ find . -iname '*.o'
./external/corefx/src/Native/Unix/System.Native/libmono_native_la-pal_io.o
./external/corefx/src/Native/Unix/System.Native/libmono_native_la-pal_tcpstate.o
./external/corefx/src/Native/Unix/System.Native/libmono_native_la-pal_maphardwaretype.o
./external/corefx/src/Native/Unix/System.Native/libmono_native_la-pal_networking.o
./external/corefx/src/Native/Unix/System.Native/libmono_native_la-pal_errno.o
./external/corefx/src/Native/Unix/System.Native/libmono_native_la-pal_networkstatistics.o
./external/corefx/src/Native/Unix/System.Native/libmono_native_la-pal_memory.o
./external/corefx/src/Native/Unix/System.Native/.libs/libmono_native_la-pal_io.o
./external/corefx/src/Native/Unix/System.Native/.libs/libmono_native_la-pal_tcpstate.o
./external/corefx/src/Native/Unix/System.Native/.libs/libmono_native_la-pal_maphardwaretype.o
./external/corefx/src/Native/Unix/System.Native/.libs/libmono_native_la-pal_networking.o
./external/corefx/src/Native/Unix/System.Native/.libs/libmono_native_la-pal_errno.o
./external/corefx/src/Native/Unix/System.Native/.libs/libmono_native_la-pal_networkstatistics.o
./external/corefx/src/Native/Unix/System.Native/.libs/libmono_native_la-pal_memory.o
./external/corefx/src/Native/Unix/System.Native/.libs/libmono_native_la-pal_random.o
./external/corefx/src/Native/Unix/System.Native/libmono_native_la-pal_random.o
```

---

Given the popularity of the [Homebrew package manager](https://brew.sh/) on macOS (particularly amongst developers), and the fact that numerous existing Homebrew packages depend on the `mono` package, it seems obviously desirable that installing mono via Homebrew should provide a fully-functional install.

This appears not to be the case at present: when mono is installed using Homebrew, `System.IO.CoreFX.FileSystemWatcher` seems to be broken.

This was reported initially as [KSP-CKAN/CKAN#2630](https://github.com/KSP-CKAN/CKAN/issues/2630), which begat [Homebrew/homebrew-core#35848](https://github.com/Homebrew/homebrew-core/issues/35848), which has unfortunately been closed because of lack of expertise amongst any of the interested parties.  It would be great if someone who understands how mono is built could assist in resolving this issue.

## Steps to Reproduce

1. Install the `ckan` homebrew package, which will cause the `mono` package to be installed if it is not already: `$ brew install ckan`
2. Run `ckan`: `$ ckan`

### Current Behaviour

The program crashes with `Unhandled Exception: System.InvalidOperationException: object_op ---> System.EntryPointNotFoundException: SystemNative_RealPath` (full stack trace below).

### Expected Behaviour

No crash, as is the case using the official .pkg:

1. Remove brewed `mono` package: `$ brew remove mono --ignore-dependencies`
2. [Download](https://www.mono-project.com/download/stable/) and install `MonoFramework-MDK-5.20.1.19.macos10.xamarin.universal.pkg`.
3. Open new Terminal window to pick up new paths.
4. Run `ckan` again: `$ ckan`
5. Observe that the error above does not occur (or at any rate, [an unrelated error](https://github.com/KSP-CKAN/CKAN/issues/2769) occurs later).

## On which platforms did you notice this

[X] macOS
[ ] Linux
[ ] Windows

**Version Used**:

Installed via Homebrew (defective):
```
Mono JIT compiler version 5.20.1.19 (tarball Thu Apr 18 12:12:03 BST 2019)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          supported, not enabled.
	Suspend:       hybrid
	GC:            sgen (concurrent by default)
```

Installed manually via downloaded .pkg (works):
```
Mono JIT compiler version 5.20.1.19 (2018-10/886c4901747 Tue Apr  9 12:37:29 EDT 2019)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          yes(600)
	Suspend:       hybrid
	GC:            sgen (concurrent by default)
```

## Stacktrace

```
Unhandled Exception:
System.InvalidOperationException: object_op ---> System.EntryPointNotFoundException: SystemNative_RealPath
  at (wrapper managed-to-native) Interop+Sys.RealPath(string)
  at System.IO.CoreFX.FileSystemWatcher+RunningInstance.Start () [0x00000] in <182506bc90704fb29de1db906cfabf73>:0 
  at System.IO.CoreFX.FileSystemWatcher.StartRaisingEvents () [0x00067] in <182506bc90704fb29de1db906cfabf73>:0 
  at System.IO.CoreFX.FileSystemWatcher.StartRaisingEventsIfNotDisposed () [0x00019] in <182506bc90704fb29de1db906cfabf73>:0 
  at System.IO.CoreFX.FileSystemWatcher.set_EnableRaisingEvents (System.Boolean value) [0x0001d] in <182506bc90704fb29de1db906cfabf73>:0 
  at (wrapper remoting-invoke-with-check) System.IO.CoreFX.FileSystemWatcher.set_EnableRaisingEvents(bool)
  at System.IO.CoreFXFileSystemWatcherProxy+<>c.<StartDispatching>b__9_0 (System.IO.CoreFX.FileSystemWatcher internal_fsw, System.IO.FileSystemWatcher fsw) [0x00048] in <182506bc90704fb29de1db906cfabf73>:0 
  at System.IO.CoreFXFileSystemWatcherProxy.Operation (System.Action`4[T1,T2,T3,T4] map_op, System.Action`2[T1,T2] object_op, System.Object handle, System.Action`2[T1,T2] cancel_op) [0x00184] in <182506bc90704fb29de1db906cfabf73>:0 
   --- End of inner exception stack trace ---
  at System.IO.CoreFXFileSystemWatcherProxy.Operation (System.Action`4[T1,T2,T3,T4] map_op, System.Action`2[T1,T2] object_op, System.Object handle, System.Action`2[T1,T2] cancel_op) [0x001a6] in <182506bc90704fb29de1db906cfabf73>:0 
  at System.IO.CoreFXFileSystemWatcherProxy.StartDispatching (System.Object handle) [0x00004] in <182506bc90704fb29de1db906cfabf73>:0 
  at System.IO.FileSystemWatcher.Start () [0x0001b] in <182506bc90704fb29de1db906cfabf73>:0 
  at System.IO.FileSystemWatcher.set_EnableRaisingEvents (System.Boolean value) [0x0003d] in <182506bc90704fb29de1db906cfabf73>:0 
  at (wrapper remoting-invoke-with-check) System.IO.FileSystemWatcher.set_EnableRaisingEvents(bool)
  at CKAN.NetFileCache..ctor (System.String path) [0x000b6] in <2c2371fb3f2949b2a6dfae930758e6ee>:0 
  at CKAN.NetFileCache..ctor (CKAN.KSPManager mgr, System.String path) [0x00000] in <2c2371fb3f2949b2a6dfae930758e6ee>:0 
  at CKAN.NetModuleCache..ctor (CKAN.KSPManager mgr, System.String path) [0x00006] in <2c2371fb3f2949b2a6dfae930758e6ee>:0 
  at CKAN.KSPManager.TrySetupCache (System.String path, System.String& failureReason) [0x0003d] in <2c2371fb3f2949b2a6dfae930758e6ee>:0 
  at CKAN.KSPManager.LoadInstancesFromRegistry () [0x000ad] in <2c2371fb3f2949b2a6dfae930758e6ee>:0 
  at CKAN.KSPManager..ctor (CKAN.IUser user, CKAN.IWin32Registry win32_registry) [0x00028] in <2c2371fb3f2949b2a6dfae930758e6ee>:0 
  at CKAN.CmdLine.MainClass.Execute (CKAN.KSPManager manager, CKAN.CmdLine.CommonOptions opts, System.String[] args) [0x0015b] in <2c2371fb3f2949b2a6dfae930758e6ee>:0 
  at CKAN.CmdLine.MainClass.Main (System.String[] args) [0x00091] in <2c2371fb3f2949b2a6dfae930758e6ee>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidOperationException: object_op ---> System.EntryPointNotFoundException: SystemNative_RealPath
  at (wrapper managed-to-native) Interop+Sys.RealPath(string)
  at System.IO.CoreFX.FileSystemWatcher+RunningInstance.Start () [0x00000] in <182506bc90704fb29de1db906cfabf73>:0 
  at System.IO.CoreFX.FileSystemWatcher.StartRaisingEvents () [0x00067] in <182506bc90704fb29de1db906cfabf73>:0 
  at System.IO.CoreFX.FileSystemWatcher.StartRaisingEventsIfNotDisposed () [0x00019] in <182506bc90704fb29de1db906cfabf73>:0 
  at System.IO.CoreFX.FileSystemWatcher.set_EnableRaisingEvents (System.Boolean value) [0x0001d] in <182506bc90704fb29de1db906cfabf73>:0 
  at (wrapper remoting-invoke-with-check) System.IO.CoreFX.FileSystemWatcher.set_EnableRaisingEvents(bool)
  at System.IO.CoreFXFileSystemWatcherProxy+<>c.<StartDispatching>b__9_0 (System.IO.CoreFX.FileSystemWatcher internal_fsw, System.IO.FileSystemWatcher fsw) [0x00048] in <182506bc90704fb29de1db906cfabf73>:0 
  at System.IO.CoreFXFileSystemWatcherProxy.Operation (System.Action`4[T1,T2,T3,T4] map_op, System.Action`2[T1,T2] object_op, System.Object handle, System.Action`2[T1,T2] cancel_op) [0x00184] in <182506bc90704fb29de1db906cfabf73>:0 
   --- End of inner exception stack trace ---
  at System.IO.CoreFXFileSystemWatcherProxy.Operation (System.Action`4[T1,T2,T3,T4] map_op, System.Action`2[T1,T2] object_op, System.Object handle, System.Action`2[T1,T2] cancel_op) [0x001a6] in <182506bc90704fb29de1db906cfabf73>:0 
  at System.IO.CoreFXFileSystemWatcherProxy.StartDispatching (System.Object handle) [0x00004] in <182506bc90704fb29de1db906cfabf73>:0 
  at System.IO.FileSystemWatcher.Start () [0x0001b] in <182506bc90704fb29de1db906cfabf73>:0 
  at System.IO.FileSystemWatcher.set_EnableRaisingEvents (System.Boolean value) [0x0003d] in <182506bc90704fb29de1db906cfabf73>:0 
  at (wrapper remoting-invoke-with-check) System.IO.FileSystemWatcher.set_EnableRaisingEvents(bool)
  at CKAN.NetFileCache..ctor (System.String path) [0x000b6] in <2c2371fb3f2949b2a6dfae930758e6ee>:0 
  at CKAN.NetFileCache..ctor (CKAN.KSPManager mgr, System.String path) [0x00000] in <2c2371fb3f2949b2a6dfae930758e6ee>:0 
  at CKAN.NetModuleCache..ctor (CKAN.KSPManager mgr, System.String path) [0x00006] in <2c2371fb3f2949b2a6dfae930758e6ee>:0 
  at CKAN.KSPManager.TrySetupCache (System.String path, System.String& failureReason) [0x0003d] in <2c2371fb3f2949b2a6dfae930758e6ee>:0 
  at CKAN.KSPManager.LoadInstancesFromRegistry () [0x000ad] in <2c2371fb3f2949b2a6dfae930758e6ee>:0 
  at CKAN.KSPManager..ctor (CKAN.IUser user, CKAN.IWin32Registry win32_registry) [0x00028] in <2c2371fb3f2949b2a6dfae930758e6ee>:0 
  at CKAN.CmdLine.MainClass.Execute (CKAN.KSPManager manager, CKAN.CmdLine.CommonOptions opts, System.String[] args) [0x0015b] in <2c2371fb3f2949b2a6dfae930758e6ee>:0 
  at CKAN.CmdLine.MainClass.Main (System.String[] args) [0x00091] in <2c2371fb3f2949b2a6dfae930758e6ee>:0 
```

## Additional information

Here is [the `mono` Homebrew formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/mono.rb).

Building the package from source using `brew install --build-from-source mono` did not make any difference, so it does not appear to be merely a defective bottle.

**Comment by jaykrell** [comment]
Could we merge System.Native into mono runtime? That would probably fix this, and make developming mono easier, like it used to be.

**Comment by lambdageek** [comment]
If I build Mono from a tarball, like Homebrew does, I see the following:

```
ld: warning: ignoring file ../../external/corefx/src/Native/Unix/System.Native/.libs/libmono_native_la-pal_io.o, file was built for unsupported file format ( 0x7F 0x45 0x4C 0x46 0x02 0x01 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 ) which is not the architecture being linked (x86_64): ../../external/corefx/src/Native/Unix/System.Native/.libs/libmono_native_la-pal_io.o
```

and indeed 

```
$ file ../../external/corefx/src/Native/Unix/System.Native/.libs/libmono_native_la-
pal_io.o
../../external/corefx/src/Native/Unix/System.Native/.libs/libmono_native_la-pal_io.o: ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), with debug_info, not stripped
```

Looks like we ship Linux object files in the tarball.

/cc @directhex

**Comment by directhex** [comment]
This is already fixed in 5.20.1.27+, and any 5.22+
